ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
+
+	return combinedNgramOffset{asc, &uni}
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
+func (a combinedNgramOffset) Get(gram ngram) simpleSection {
+	if a.asc == nil {
+		return simpleSection{}
+	}
+
+	var sec simpleSection
+	if gram&ngramAsciiMask == gram {
+		sec = a.asc.Get(gram)
+	}
+	if sec.sz == 0 || sec.sz == ngramAsciiMaxSectionLength {
+		sec = a.uni.Get(gram)
+	}
+	return sec
+}
+
+func (a combinedNgramOffset) DumpMap() map[ngram]simpleSection {
+	m := a.asc.DumpMap()
+	for k, v := range a.uni.DumpMap() {
+		m[k] = v
+	}
+	return m
+}
+
+func (a combinedNgramOffset) SizeBytes() int {
+	return a.asc.SizeBytes() + a.uni.SizeBytes()
+}
+
+const ngramAsciiMask = 127 | 127<<21 | 127<<42
+
+// Ascii mapping packs 3*7b chars and 11 bits of lengths, with this as the set maximum.
+// We could save another ~3% of total RAM / 5% of combinedNgramOffset RAM by switching to
+// a 40b packing with 19-bit lengths, but the code would be significantly uglier so it doesn't
+// seem worth it.
+const ngramAsciiMaxSectionLength = (1 << 11) - 1
+
+func ngramAsciiToPacked(ng ngram) uint32 {
+	return uint32(ng&127) | uint32((ng>>(21-7))&(127<<7)) | uint32((ng>>(42-14))&(127<<14))
+}
+
+func ngramAsciiPackedToNgram(ng uint32) ngram {
+	return ngram(ng&127) | ngram(ng&(127<<7))<<(21-7) | ngram(ng&(127<<14))<<(42-14)
+}
+
+// asciiNgramOffset stores ascii trigrams packed together with short lengths,
+// using offsets for a chunk of entries to limit the number of lengths that must
+// be summed to compute a section's offset.
+type asciiNgramOffset struct {
+	entries      []uint32 // (chara << 25 | charb << 18 | charc << 11 | length)
+	chunkOffsets []uint32 // offset for entries[i*asciiNgramOffsetChunkLength]
+}
+
+// asciiNgramOffsetChunkLength specifies how many entries share one initial offset.
+// It must be a power of 2, and was chosen empirically by measuring RAM usage:
+// 8: 4132MB, 16: 4047MB, 32: 4006MB, 64: 3992MB, 128: 3990MB
+const asciiNgramOffsetChunkLength = 32
+
+func makeAsciiNgramOffset(ngrams []uint32, offsets []uint32) *asciiNgramOffset {
+	// fmt.Println("makeAsciiNgramOffset", ngrams, offsets)
+	ao := &asciiNgramOffset{
+		entries:      make([]uint32, 0, len(ngrams)),
+		chunkOffsets: make([]uint32, 0, len(ngrams)/asciiNgramOffsetChunkLength),
+	}
+
+	for i, ng := range ngrams {
+		if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+			ao.chunkOffsets = append(ao.chunkOffsets, offsets[i])
+		}
+		length := offsets[i+1] - offsets[i]
+		for length > 0 {
+			if length >= ngramAsciiMaxSectionLength {
+				ao.entries = append(ao.entries, ng<<11|ngramAsciiMaxSectionLength)
+				length -= ngramAsciiMaxSectionLength
+			} else {
+				ao.entries = append(ao.entries, ng<<11|length)
+				length = 0
+			}
+			if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+				break
+			}
+		}
```

Comment:
This body needs more explanation. It took a bunch of scrolling around for me to understand why you add multiple entries until you fill up the "chunk".

Correct me if I am wrong: even though we can't encode lengths bigger than `ngramAsciiMaxSectionLength`, we need the lengths to correctly calculate future offsets in the chunk. So we split the lengths up in the chunk such that they add up, so that when we do the offset calculation in Get we get back the original value. However, if we fill up the chunk while doing this we can stop the encoding process for this since the next value will be a fresh chunk with an encoded offset in `chunkOffsets`.

Reply:
Added comments to explain the reasoning and rearranged the loops to hopefully be clearer.